Documentation Overhaul (#204)
projecttacoma/fqm-execution: README.md

- `options` <[CalculatorTypes.CalculationOptions](https://github.com/projecttacoma/fqm-execution/blob/7c43b94521963703bdb43932bad60e5be3e7eeaa/src/types/Calculator.ts#L12)>

**Returns**:

[Promise<CalculatorTypes.ValueSetOutput>](https://github.com/projecttacoma/fqm-execution/blob/7c43b94521963703bdb43932bad60e5be3e7eeaa/src/types/Calculator.ts#L404)

## Calculation Options

The options that we support for calculation are as follows:

- `[calculateClauseCoverage]`<[boolean]()>: Include HTML structure with clause coverage highlighting (default: `true`)
- `[calculateHTML]`<[boolean]()>: Include HTML structure for highlighting (default: `true`)
- `[calculateSDEs]`<[boolean]()>: Include Supplemental Data Elements (SDEs) in calculation (default: `true`)
- `[clearElmJsonsCache]`<[boolean]()>: if `true`, clears ELM JSON cache before running calculation (default: `false`)
- `[enableDebugOutput]`<[boolean]()>: Enable debug output including CQL, ELM, results (default: `false`)
- `[includeClauseResults]` <[boolean]()>: Option to include clause results (default: `false`)
- `[measurementPeriodEnd]`<[string]()>: End of measurement period

> - `[measurementPeriodStart]`<[string]()>: Start of measurement period

Reviewer: Do we want to add a little blurb here about how these being optional works?

Author: yes, good call

npm install -g fqm-execution
```

# Usage

## Quickstart

```typescript
import { Calculator } from 'fqm-execution';

// ...

const { results } = await Calculator.calculate(measureBundle, patientBundles, options[, valueSetCache]);

// ...
```

## Requirements

To calculate a FHIR-based eCQM, `fqm-execution` needs the following information:

- A [FHIR Bundle](http://hl7.org/fhir/bundle.html) resource that contains:
  - One [FHIR Measure](https://build.fhir.org/ig/HL7/cqf-measures/StructureDefinition-measure-cqfm.html) resource containing the metadata for the measure
  - The [FHIR Library](https://build.fhir.org/ig/HL7/cqf-measures/StructureDefinition-library-cqfm.html) resource that is referenced from the above `Measure` resource's `.library` property, and any other FHIR `Library` resources that are dependencies of the main measure logic (e.g. `FHIRHelpers`)
    - **NOTE**: These `Library` resources _must_ contain `base64`-encoded ELM JSON content directly on the resource. `fqm-execution` does not do any real-time translation of CQL to ELM
  - Any [FHIR ValueSet](http://hl7.org/fhir/valueset.html) resources that are used in the measure logic\*

> - One or more FHIR `Bundle` resources that contain:

Reviewer 2: Should we specify the instances in which we do not need patient bundles? Like dataRequirements and queryInfo? Or is this just for calculate

Author: The intent was the latter since it's under a sentence that says "to calculate a FHIR-based eCQM" (most common use case)

I'm gonna add an aside anyway though

  - One [FHIR Patient](http://hl7.org/fhir/patient.html) resource
  - Any other FHIR resources that contain relevant data for the above patient that should be considered during measure calculation

\*`ValueSet` resources can be omitted if you follow the approach outlined in the [ValueSet Resolution](#valueset-resolution) section

## Calculating an eCQM

`fqm-execution` exports the `Calculator` object, which contains an API function that returns eCQM calculation results as a detailed object containing calculation results for each patient across all population groups defined in `Measure` resource:

```typescript
import { Calculator } from 'fqm-execution';

// ...

const { results } = await Calculator.calculate(measureBundle, patientBundles, options[, valueSetCache]);

> /*        ^?

Author: For this first one, yeah it makes sense to separate since it's the whole structure. For the smaller examples later on, I think it's fine to leave as is since people who are in those sections likely have more exposure to results already

```typescript
[
  {
    patientId: 'patient-1',
    detailedResults: [
      {
        groupId: 'group-1',
        populationResults: [
          {
            populationType: 'initial-population',
            criteriaExpression: 'Initial Population',
            result: true
          },
          {
            populationType: 'denominator',
            criteriaExpression: 'Denominator',
            result: true
          },
          {
            populationType: 'numerator',
            criteriaExpression: 'Numerator',
            result: true
          }
        ]
        // ...
      }
      // ...
    ]
    // ...
  }
];
```

These results are assembled according to the various [Measure Population Semantics](https://build.fhir.org/ig/HL7/cqf-measures/measure-conformance.html#measure-population-semantics) defined in the FHIR Quality Measure Implementation Guide, meaning that dependencies of the various eCQM populations are factored in
during the results processing (e.g. a patient cannot be in the numerator if they are not in the denominator).

### Episode-based Measures

> For episode-based measures, calculation is largely the same as the [basic example](#interpreting-calculation-results), but the detailed results responses allows for a more granular view of which populations each individual episode landed into via an `episodeResults` list that appears in each

Author: My high school English teacher just cringed for me not recognizing the subject of the sentence here

### `.calculateGapsInCare`

Get gaps in care for each patient. Output adheres to the [DEQM Gaps In Care Bundle Profile](https://build.fhir.org/ig/HL7/davinci-deqm/StructureDefinition-gaps-bundle-deqm.html)

**Parameters**:

- `measureBundle` <[fhir4.Bundle](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5f88f0c7a11e86a19fc2356d0e11dfa7f472e3c9/types/fhir/r4.d.ts#L4325)>
- `patientBundles` <[fhir4.Bundle[]](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5f88f0c7a11e86a19fc2356d0e11dfa7f472e3c9/types/fhir/r4.d.ts#L4325)>
- `options` <[CalculatorTypes.CalculationOptions](https://github.com/projecttacoma/fqm-execution/blob/7c43b94521963703bdb43932bad60e5be3e7eeaa/src/types/Calculator.ts#L12)>
- `[valueSetCache]` <[fhir4.ValueSet[]](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5f88f0c7a11e86a19fc2356d0e11dfa7f472e3c9/types/fhir/r4.d.ts#L28217)>

**Returns**:

[Promise<CalculatorTypes.GICCalculationOutput>](https://github.com/projecttacoma/fqm-execution/blob/7c43b94521963703bdb43932bad60e5be3e7eeaa/src/types/Calculator.ts#L397)

### `.calculateLibraryDataRequirements`

> Get data requirements for a given Library bundle with a root Library reference

Reviewer 2: ```suggestion
Get data requirements for a given Library bundle with a root Library reference.
```

## Measure Bundle Helpers

```typescript
import { MeasureBundleHelpers } from 'fqm-execution';
```

### `.addValueSetsToMeasureBundle`

Add missing ValueSet resources to a measure bundle.

**Parameters**:

- `measureBundle` <[fhir4.Bundle](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5f88f0c7a11e86a19fc2356d0e11dfa7f472e3c9/types/fhir/r4.d.ts#L4325)>
- `options` <[CalculatorTypes.CalculationOptions](https://github.com/projecttacoma/fqm-execution/blob/7c43b94521963703bdb43932bad60e5be3e7eeaa/src/types/Calculator.ts#L12)>

**Returns**:

[Promise<CalculatorTypes.ValueSetOutput>](https://github.com/projecttacoma/fqm-execution/blob/7c43b94521963703bdb43932bad60e5be3e7eeaa/src/types/Calculator.ts#L404)

## Calculation Options

The options that we support for calculation are as follows:

- `[calculateClauseCoverage]`<[boolean]()>: Include HTML structure with clause coverage highlighting (default: `true`)
- `[calculateHTML]`<[boolean]()>: Include HTML structure for highlighting (default: `true`)
- `[calculateSDEs]`<[boolean]()>: Include Supplemental Data Elements (SDEs) in calculation (default: `true`)

> - `[clearElmJsonsCache]`<[boolean]()>: if `true`, clears ELM JSON cache before running calculation (default: `false`)

Reviewer 2: ```suggestion
- `[clearElmJsonsCache]`<[boolean]()>: If `true`, clears ELM JSON cache before running calculation (default: `false`)
```

- `[enableDebugOutput]`<[boolean]()>: Enable debug output including CQL, ELM, results (default: `false`)
- `[includeClauseResults]` <[boolean]()>: Option to include clause results (default: `false`)
- `[measurementPeriodEnd]`<[string]()>: End of measurement period. Defaults to the `.effectivePeriod.end` on the `Measure` resource, but can be overridden or specified using this option, which will take precedence

> - `[measurementPeriodStart]`<[string]()>: Start of measurement period. Defaults to the `.effectivePeriod.start` on the `Measure` resource, but can be overridden or specified using this option, which will take precedence

Reviewer 3: The descriptions here for measurement period start/end look good. Should we replace the current descriptions under "CLI Options" with these descriptions? (The current "CLI Options" descriptions still mention the defaulting to 2019 (CLI options weren't edited so they're not in the code diff))

- `[patientSource]`<[DataProvider](https://github.com/cqframework/cql-execution/blob/e4d3f24571daf3ae9f891df11bb22fc964f6de5d/src/types/cql-patient.interfaces.ts#L8)>: PatientSource to use. **If provided, the `patientBundles` argument must be `[]`**. See the [Custom PatientSource section](#custom-patientsource) for more info

> - `[reportType]`<['individual' | 'subject-list' | 'summary']()>: The type of FHIR MeasureReport to return (default: `'individual'`)

Reviewer 3: Should we mention that `subject-list` is not fully supported at this time?

Author: What if I just..... removed it from the docs until it's supported ;)

Reviewer 3: What a good idea... I like the way you think mgramigna 🧠

- `[returnELM]`<[boolean]()>: Enables the return of ELM Libraries and name of main library to be used for further processing (e.g. gaps in care) (default: `false`)
- `[rootLibRef]`<[string]()>: Reference to root library to be used in `calculateLibraryDataRequirements`. Should be a canonical URL but resource ID will work if matching one exists in the bundle

> - `[trustMetaProfile]`<[boolean]()>: If `true` trust the content of `meta.profile` as a source of truth for what profiles the data that `cql-exec-fhir` grabs validates against. **Use of this option will cause `cql-exec-fhir` to filter out resources that don't have a valid `meta.profile` attribute** (default: `false`)

Reviewer 2: ```suggestion
- `[trustMetaProfile]`<[boolean]()>: If `true`, trust the content of `meta.profile` as a source of truth for what profiles the data that `cql-exec-fhir` grabs validates against. **Use of this option will cause `cql-exec-fhir` to filter out resources that don't have a valid `meta.profile` attribute** (default: `false`)
```

Options:
  --debug                                     Enable debug output. (default: false)
  --slim                                      Use slimmed-down calculation results interfaces (default: false)
  --report-type <report-type>                 Type of report, "individual", "summary"
  -m, --measure-bundle <measure-bundle>       Path to measure bundle.

>   -p, --patient-bundles <patient-bundles...>  Paths to patient bundles. Required unless output type is one of the following: dataRequirements, queryInfo, valueSets.

Reviewer 2: ```suggestion
  -p, --patient-bundles <patient-bundles...>  Paths to patient bundles. Required unless output type is one of the following: dataRequirements, libraryDataRequirements, queryInfo, valueSets.
```

E.g.

```bash
# Generate detailed calculation results by calculating a measure on a patient bundle:
fqm-execution detailed -m /path/to/measure/bundle.json -p /path/to/patient/bundle.json -o detailed-results.json

# Generate a MeasureReport by calculating a measure on multiple patient bundles:
fqm-execution reports -m /path/to/measure/bundle.json -p /path/to/patient1/bundle.json /path/to/patient2/bundle.json -o reports.json
```

# Guides and Concepts

## Measures with Observation Functions

> For [Continuous Variable Measures](https://build.fhir.org/ig/HL7/cqf-measures/measure-conformance.html#continuous-variable-measure) and [Ratio Measures](https://build.fhir.org/ig/HL7/cqf-measures/measure-conformance.html#ratio-measures), some of the measure populations can have an associates "measure observation", which is a CQL function that will return some result based on some information present on the data during calculation. In the case of these measures, `fqm-execution` will include an `observations` property on the `populationResult` object for each measure observation population. This `observations` property

Author: Dangit!

                criteriaExpression: 'daysObservation',
                result: true,
                observations: [2]
              }
            ]
          }
        ]
      }
    ]
  }
];
```

From these results, it can be seen that `"episode-1"` has a duration of `1` day, and `"episode-2"` has a duration of `2` days. The results are also indicating the `populationId` property on each `populationResult`, as well as the `criteriaReferenceId` property on each measure observation `populationResult`. In the case of measure observations, it is common to assign an `id` property to each population in the `Measure` resource, as well as use the [cqfm-criteriaReference extension](https://build.fhir.org/ig/HL7/cqf-measures/StructureDefinition-cqfm-criteriaReference.html) on the measure observation population in order to indicate which population that function is observing (denominator in this example). Doing so allows consumers of the above result to easily identify which populations the measure observation results are referring to, which is especially useful if there are multiple measure observations defined.

## `meta.profile` Checking

> `fqm-execution` can be configured to use a [trusted environment with cql-exec-fhir](https://build.fhir.org/ig/HL7/cqf-measures/StructureDefinition-cqfm-criteriaReference.html), where resources returned as the result of an ELM Retrieve expressions will only be included if they have a `meta.profile` list that contains the canonical URL of the profile being asked for by the Retrieve. To enable this environment, use the `trustMetaProfile` calculation option during calculation:

Reviewer: ```suggestion
`fqm-execution` can be configured to use a [trusted environment with cql-exec-fhir](https://build.fhir.org/ig/HL7/cqf-measures/StructureDefinition-cqfm-criteriaReference.html), where resources returned as the result of an ELM Retrieve expression will only be included if they have a `meta.profile` list that contains the canonical URL of the profile being asked for by the Retrieve. To enable this environment, use the `trustMetaProfile` calculation option during calculation:
```

Author: DANGIT

                criteriaExpression: 'daysObservation',
                result: true,
                observations: [2]
              }
            ]
          }
        ]
      }
    ]
  }
];
```

From these results, it can be seen that `"episode-1"` has a duration of `1` day, and `"episode-2"` has a duration of `2` days. The results are also indicating the `populationId` property on each `populationResult`, as well as the `criteriaReferenceId` property on each measure observation `populationResult`. In the case of measure observations, it is common to assign an `id` property to each population in the `Measure` resource, as well as use the [cqfm-criteriaReference extension](https://build.fhir.org/ig/HL7/cqf-measures/StructureDefinition-cqfm-criteriaReference.html) on the measure observation population in order to indicate which population that function is observing (denominator in this example). Doing so allows consumers of the above result to easily identify which populations the measure observation results are referring to, which is especially useful if there are multiple measure observations defined.

## `meta.profile` Checking

> `fqm-execution` can be configured to use a [trusted environment with cql-exec-fhir](https://build.fhir.org/ig/HL7/cqf-measures/StructureDefinition-cqfm-criteriaReference.html), where resources returned as the result of an ELM Retrieve expression will only be included if they have a `meta.profile` list that contains the canonical URL of the profile being asked for by the Retrieve. To enable this environment, use the `trustMetaProfile` calculation option during calculation:

Reviewer 3: I think the link here might be incorrect (links to the criteriaReference extension that is referenced in the previous section)

Author: Nah this was intentional. It's an exercise to the reader to find the correct link

});

// results
/*    ^?
      [
          {
            "patientId": "patient-1",
            "detailedResults": [
              {
                "groupId": "group-1",
                "html": ""<div><h2>Population Group: group-1</h2> ..."
                ...
              }
              ...
            ]
            ...
          }
          ...
     ]
*/
```

In `MeasureReport` calculation, an HTML string will appear on the `.text.div` property of the returned FHIR `MeasureReport` for a given patient:

```typescript
import { Calculator } from 'fqm-execution';

// ...

const { results } = await Calculator.calculateMeasureReports(measureBundle, patientBundles, {
  calculateHTML: true /* true by default, but explicitly shown for demonstration */,
  reportType: 'individual' /* highlighting is only available for individual MeasureReport calculation */
});

// results
/*    ^?
        [
          {
            "id": "<some-random-uuid>",
            "resourceType": "MeasureReport",
            "text": {
              "status": "generated",
              "div": "<div><h2>Population Group: group-1</h2> ..."
            }
            ...
          }
          ...
        ]
 */
```

> **NOTE**: Measure logic highlighting is not available when calculation a population `MeasureReport` with `reportType = 'summary'`

Reviewer 3: ```suggestion
**NOTE**: Measure logic highlighting is not available when calculating a population `MeasureReport` with `reportType = 'summary'`
```

Author: These non-typo wrong word errors will be the end of me

## Group Clause Coverage Highlighting

> `fqm-execution` can generated highlighted HTML that indicates which individual pieces of the measure logic CQL were processed at all during calculation, regardless of whether or not they held "truthy" or "falsy" value. This is often referred to as "Clause Coverage". "Covered" clauses will

Reviewer 2: ```suggestion
`fqm-execution` can generate highlighted HTML that indicates which individual pieces of the measure logic CQL were processed at all during calculation, regardless of whether or not they held "truthy" or "falsy" values. This is often referred to as "Clause Coverage". "Covered" clauses will
```

Author: You could tell I was losing steam at this point

Reviewer 2: For the amount of text in this PR, the lack of grammar/spelling errors is genuinely baffling.

## Custom PatientSource

Via the `patientSource` calculation option, users can provide their own provider of patient data to `fqm-execution` that will be called by the underlying `cql-execution` engine. The [type definitions in cql-execution](https://github.com/cqframework/cql-execution/blob/e4d3f24571daf3ae9f891df11bb22fc964f6de5d/src/types/cql-patient.interfaces.ts#L8) indicate which functions a custom `DataProvider` must implement
to work alongside `cql-execution`. The following skeleton code showcases the implementations that would be required for a hypothetical custom `DataProvider`:

```typescript
import { PatientObject, DataProvider, RecordObject } from 'cql-execution';

// NOTE: it is recommended here to re-use the FHIRObject class from cql-exec-fhir, if possible
class CustomRecordObject implements RecordObject {
  get(field: any) {
    /* Implementation that returns the value of any field on this object */
  }

  // Required by cql-execution API
  getId() {
    /* Implementation that returns the value of the id field on this object */
  }

  // Required by cql-execution API (but not currently used in the cql-exec-fhir FHIR data model)
  getCode(field: any) {

>     /* Implementation that returns the value the field on this object as a cql-execution Code type */

Reviewer 2: ```suggestion
    /* Implementation that returns the value of the field on this object as a cql-execution Code type */
```
Not sure if this is what you meant, maybe something else would be better

  }

  // Required by cql-execution API (but not currently used in the cql-exec-fhir FHIR data model)
  getDate(field: any) {

>     /* Implementation that returns the value the field on this object as a cql-execution Date type */

Reviewer 2: ```suggestion
    /* Implementation that returns the value of the field on this object as a cql-execution Date type */
```

  }

  // Required by cql-execution API (but not currently used in the cql-exec-fhir FHIR data model)
  getDateOrInterval(field: any) {

>     /* Implementation that returns the value the field on this object as a cql-execution Date or Interval type */

Reviewer 2: ```suggestion
    /* Implementation that returns the value of the field on this object as a cql-execution Date or Interval type */
```

  // ...
}

// Check if some measure scoring code is the code for ratio measure scoring
if (someScoringCode === Enums.MeasureScoreType.RATIO) {
  // ...
}

// Check if some population result has a truthy final result
if (somePopulationFinalResult === Enums.FinalResult.TRUE) {
  // ...
}

// Check if some population result is not relevant for eCQM calculation (i.e. wasn't considered)
if (somePopulationRelevanceResult === Enums.Relevance.NA) {
  // ...
}
```

- `CQLTypes`: Low-level types that interface with the `cql-execution` engine. Unlikely to be needed outside of `fqm-execution` itself, but provided anyway for convenience
- `ELMTypes`: Low-level types that define basic structure of ELM. Does not define the entire ELM logical specification. Unlikely to be needed outside of `fqm-execution` itself, but provided anyway for convenience

# Recipes

## Displaying Highlighted HTML in a React App

A common use case for the highlighted logic HTML described in the [Measure Logic Highlighting section](#measure-logic-highlighting) is displaying it in a web application of some kind. In the case of React trying to render an HTML string directly will not work

> to due React's security measures to prevent Cross Site Scripting by escaping/sanitizing HTML strings that attempt to be rendered in the DOM.

Reviewer 2: ```suggestion
due to React's security measures to prevent Cross Site Scripting by escaping/sanitizing HTML strings that attempt to be rendered in the DOM.
```

Author: This one is embarassing
